[export] add _union_dataclass to support comparing dataclasses that inherits from union.

ydwu4 opened the pull request:
Summary: We want to share the __eq__ and __repr__ method of schema dataclasses that inherits from _Union. Added _union_dataclass to support this.

Test Plan: added new test.

Differential Revision: D72068295

Skylion007 commented on torch/_export/serde/union.py:
```diff
@@ -32,6 +32,17 @@ def _get_field_names(cls) -> set[str]:
+# For example, _union_dataclass customizes the __eq__ method to only check the type
+# and value property instead of default implmentation of dataclass which goes
+# through every field in the dataclass.
+def _union_dataclass(cls):
```
```suggestion
@dataclass_transform()
def _union_dataclass(cls: T) -> T:

Skylion007 replied:
Double check the args to dataclass_transform

Skylion007 commented on torch/_export/serde/union.py:
```diff
@@ -1,7 +1,7 @@
 import functools
 from collections.abc import Hashable
-from dataclasses import fields
+from dataclasses import dataclass, fields
```
from dataclasses import dataclass, fields
from typing_extensions import dataclass_transform

ydwu4 committed:
[export] add _union_dataclass to support comparing dataclasses that inherits from union. (#155932)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/155932

We want to share the __eq__ and __repr__ method of schema dataclasses that inherits from _Union. Added _union_dataclass to support this.

Test Plan: added new test.

Reviewed By: angelayi

Differential Revision: D72068295
